backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java

```diff
@@ -153,4 +168,149 @@ public void resetDemoStudents() {
+        }
+    }
+
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public ExternalUserDto createExternalUser(Integer courseExecutionId, ExternalUserDto externalUserDto) {
+        verifyEmail(externalUserDto);
+        verifyRole(externalUserDto);
+        CourseExecution courseExecution = getCourseExecution(courseExecutionId);
+        User user1 = getUser(externalUserDto, courseExecution);
+        associateUserWithExecution(courseExecution, user1);
+        sendConfirmationEmailTo(user1);
+        return new ExternalUserDto(user1);
+    }
+
+    public String generateConfirmationToken(User user) {
+        String token = KeyGenerators.string().generateKey();
+        user.setTokenGenerationDate(LocalDateTime.now());
+        user.setConfirmationToken(token);
+        return token;
+    }
+
+    private void sendConfirmationEmailTo(User user) {
+        generateConfirmationToken(user);
+        mailer.sendSimpleMail(Mailer.MAIL_USER, user.getEmail(), User.PASSWORD_CONFIRMATION_MAIL_SUBJECT, buildMailBody(user));
+    }
+
+    private String buildMailBody(User user) {
+        String msg = "To confirm your registration click the following link";
+        return String.format("%s: %s", msg, LinkHandler.createConfirmRegistrationLink(user.getEmail(), user.getConfirmationToken()));
+    }
+
+    private void associateUserWithExecution(CourseExecution courseExecution, User user1) {
+        courseExecution.addUser(user1);
+        user1.addCourse(courseExecution);
+    }
+
+    private User getUser(ExternalUserDto externalUserDto, CourseExecution courseExecution) {
+        Optional<User> user = userRepository.findByUsername(externalUserDto.getEmail());
+        User user1;
+        if(user.isPresent()){
+            user.get().addCourse(courseExecution);
+            user1 = user.get();
+        }else{
+            user1 = new User("", externalUserDto.getEmail(), externalUserDto.getRole());
+            userRepository.save(user1);
+            user1.setState(User.State.INACTIVE);
+        }
+        user1.setAdmin(false);
+        user1.setEmail(externalUserDto.getEmail());
+        return user1;
+    }
+
+    private CourseExecution getCourseExecution(Integer courseExecutionId) {
+        CourseExecution courseExecution = courseExecutionRepository.findById(courseExecutionId)
+                .orElseThrow(() -> new TutorException(COURSE_EXECUTION_NOT_FOUND, courseExecutionId));
+
+        if (courseExecution.getType() != Course.Type.EXTERNAL){
+            throw new TutorException(COURSE_EXECUTION_NOT_EXTERNAL, courseExecutionId);
+        }
+        return courseExecution;
+    }
+
+    private void verifyRole(ExternalUserDto externalUserDto) {
+        if(externalUserDto.getRole() == null)
+            throw new TutorException(INVALID_ROLE);
+    }
+
+    private void verifyEmail(ExternalUserDto externalUserDto) {
+        if(externalUserDto.getEmail() == null || externalUserDto.getEmail().trim().equals(""))
```

Comment:
better in user, and use a regexp to validate format, e.g. https://blog.mailtrap.io/java-email-validation/